### ISSUE TITLE
Publish as conda package

### ISSUE THREAD

**Issue by benjello** [openfisca/openfisca-core#1039]
Hi there!

I really enjoy OpenFisca, but I recently encountered an issue.

### Here is what I did:

I want to advocate openfisca to non-dev (mainly economists) Windows users relying on (standard) Anaconda for Python stuff.
 
### Here is what I expected to happen:

Install the conda package at least for openfisca-core.
Install using pip breaks because missing windows libraries.  

### Here is what actually happened:

I solved the problem by installing using conda as much packages (numpy etc)  as i could then pip installing the openfisca-core package.
 
It would be nice to be able to install openfisca-core using conda.

## Context

I identify more as a:

- Analyst _(I make macroscopic computations on real populations)_.
- Business expert _(I create tests and model legislation)_.
- Developer _(I create tools that use the existing OpenFisca code)_.

**Comment by MattiSG** [comment]
Thanks @benjello for creating this issue! I also encountered this issue when installing on an M1 setup, and had to fix it with #990.I'm not sure I fully understand what creating a conda package entails, in terms of maintenance, technical consequences, and benefits for reusers. Do we have someone here with enough experience to make a recap? 😃

**Comment by sandcha** [comment]
I also have never used conda but from the [Building conda packages from scratch page](https://docs.conda.io/projects/conda-build/en/latest/user-guide/tutorials/build-pkgs.html), creating a package looks like creating 3 files (`meta.yaml`, `bld.bat` and `build.sh`) that would use our `setup.py` to get the package dependencies and could target Pypi (or our GitHub repositories).
 
If this description isn't missing some step, the 3 files shouldn't need much maintenance and [uploading the package to anaconda.org](https://docs.conda.io/projects/conda-build/en/latest/user-guide/tutorials/build-pkgs.html#optional-uploading-new-packages-to-anaconda-org) could be done by one command (+ install of this command + account on Anaconda.org). So, it looks like a step that we can add to the CI (and I agree that it should be integrated to the CI if we are looking for a sustainable solution).

As for the benefits for the users/economists on Windows, from what I understand, it would mean bringing openfisca to their usual environment instead of making them adapt the developers way of getting openfisca on another platform with all the mishaps that it comes with. And, from a perspective where installing a new version regularly is already painful and comes before any "real work", having to adapt is even a barrier to consider using openfisca.

So, I would say that it's time to find a non exotic solution for users on Windows. 
Conda is one OpenFisca them. [Poetry](https://python-poetry.org/docs/#system-requirements) might be another option 🤔

**Comment by benjello** [comment]
I would favor conda over Poetry for the kind of user I target.
It worked pretty well with my students. I asked them to install conda create ann env with a few important packages that requires Windows specific libs (numpy pandas pytables etc). 
Since I only used openfisca within notebooks, the openfisca packages were installed by pip within those notebooks.
 
Having a conda package would definitely ease the first steps.

I tried Poetry and I still have problem using it ...

**Comment by benoit-cty** [comment]
Hello,

In addition, don't having Openfisca on Conda prevent others to push packages that rely on openfisca on conda.
I'm a contributor of CodeCarbon where we use github action to push on Conda.
And for Leximpact I use NBDev that have a system to push to conda using the pip package.

I could have a look at this issue.

**Comment by benoit-cty** [comment]
I achieve to upload OpenFisca-Core on Conda-Forge : https://anaconda.org/leximpact/openfisca-core

What I've done:
- Add `conda/meta.yaml` in the project
- `docker run --rm -t -i -v $PWD:/src continuumio/miniconda3 /bin/bash`
  - `cd /src`
  - `conda install conda-build`
  - `conda build -c conda-forge --output-folder ./conda-out/ ./conda/`
  - `conda install anaconda-client`
  - `anaconda login`
  - `anaconda upload ./conda-out/noarch/openfisca-core-0.0.1-py_0.tar.bz2`

Please note that:
- Unlike PiPy, the package name must be lowercase.
- dpath >=1.5.0,<2.0.0' don't exist in conda-forge, only 1.4 or >2.0.1, so I use 
- It is an early work, I have to look at different options to add it to the CI

Tutorials used:
- https://medium.com/analytics-vidhya/publish-a-python-package-to-conda-b352eb0bcb2e

What to do next:

- [ ] Create a PR: https://github.com/openfisca/openfisca-core/pull/1099
- [ ] Create an OpenFisca account on Conda Forge, to use it instead of LexImpact.
- [ ] For conda-forge there is a way to use the pypi package source: https://blog.gishub.org/how-to-publish-a-python-package-on-conda-forge
- [ ] Add the build to the CI
- [ ] Use Jinja to use env variable as version https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#templating-with-jinja

**Comment by benoit-cty** [comment]
Finaly there is no need to modify 

I achieve to upload OpenFisca-Core on Anaconda : https://anaconda.org/leximpact/openfisca-core

#### New features

- Introduce a way to add OpenFisca to [Anaconda](https://www.anaconda.com/)

#### Technical changes

- Add conda/meta.yaml
- Add CI to publish to [Conda Forge](https://conda-forge.org/)



#### What I've done so far
- Add `conda/meta.yaml` in the project
- `docker run --rm -t -i -v $PWD:/src continuumio/miniconda3 /bin/bash`
  - `cd /src`
  - `conda install conda-build`
  - `conda build -c conda-forge --output-folder ./conda-out/ ./conda/`
  - `conda install anaconda-client`
  - `anaconda login`
  - `anaconda upload ./conda-out/noarch/openfisca-core-0.0.1-py_0.tar.bz2`

#### Please note that
- Unlike PiPy, the package name must be lowercase.
- dpath >=1.5.0,<2.0.0' don't exist in conda-forge, only 1.4 or >2.0.1, so I use 
- It is an early work, I have to look at different options to add it to the CI

Tutorials used:
- https://medium.com/analytics-vidhya/publish-a-python-package-to-conda-b352eb0bcb2e

BUT this easy way publish the package in a specific "channel" of Anaconda, to be on conda-forge, we need to use another procedure:
For publishing to Conda-Forge, we have to fork a project and made a PR back to it. There is automated build in Linux, OSX and Windows: https://github.com/conda-forge/staged-recipes/pull/17399
 - "Propose the change as a pull request. Your recipe will automatically be built on Windows, Linux, and OSX to test that it works, but the distribution will not yet be available on the conda-forge channel."
- "Once the recipe is ready it will be merged and a new "feedstock" repository will automatically be created for the recipe. The build and upload processes take place in the feedstock, and once complete the package will be available on the conda-forge channel."

A chat is available for conda-forge integration at https://gitter.im/conda-forge/conda-forge.github.io

#### TODO

- [x] Create a PR: https://github.com/openfisca/openfisca-core/pull/1099
- [x] Create an OpenFisca account on Conda Forge, to use it instead of LexImpact. => Not needed for feedstock ?
- [ ] Upgrade dpath : https://github.com/openfisca/openfisca-core/pull/1065
- [ ] Make a new repo https://github.com/conda-forge/staged-recipes/pull/17399
- [ ] For conda-forge there is a way to use the pypi package source: https://blog.gishub.org/how-to-publish-a-python-package-on-conda-forge
- [ ] Add the build to the CI
- [ ] Use Jinja to use env variable as version https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#templating-with-jinja

**Comment by benoit-cty** [comment]
My PR to add openfisca-core is merged, so there is now an openfisca-core teams in Conda-Forge https://github.com/orgs/conda-forge/teams/openfisca-core
And, we now have OpenFisca-Core in Conda forge : https://anaconda.org/conda-forge/openfisca-core :smiley:

**Comment by benoit-cty** [comment]
Bad news : OpenFisca-France can't be pushed to conda-forge due to long filename : the Windows CI crash on that.
And it even can't be build locally to push it on conda openfisca channel : `error: could not create 'C:\Users\ben\anaconda3\conda-bld\openfisca-france_1641913882030\_h_env\Lib\site-packages\openfisca_france\parameters\chomage\allocations_assurance_chomage\alloc_unique_degressive\coefficient_degressivite\affiliation_14_sur_24_derniers_mois\entre_25_50_ans.yaml': No such file or directory` : The path is 261 char above 255 Windows limit. This is the default path used by Anaconda.  The actual limit of 150 must be lowered to 120 to succeed the CI.@Sasha-Laniece has looked how to reduce filepath but it will be too hard : 360 parameters concerned.

TODO for OF-Core:
- [ ] Handle optional installation
- [ ] Add other maintainer
- [ ] Document the publishing process : 
- [ ] Update the OF-Core README to add Conda usage

**Comment by benjello** [comment]
May be we should look for long file names again. Try to shorten the more complicated first nodes and simplify others.

**Comment by benoit-cty** [comment]
I achieve to push openfisca-france to conda: https://anaconda.org/OpenFisca.org/openfisca-france
It's not conda-forge but it works on my Windows machine with `conda install -c openfisca.org openfisca-france`
What I've done : https://github.com/openfisca/openfisca-france/blob/publish-to-conda/.conda/README.md

**Comment by benoit-cty** [comment]
I've opened an issue to the conda team for long pathname : https://github.com/conda-forge/staged-recipes/issues/17519

And begin a PR to discuss OpenFisca-France CI: https://github.com/openfisca/openfisca-france/pull/1786

**Comment by sandcha** [comment]
The first openfisca-core publishing raises a question about the needed `extra-require` dependencies. 
In order to identify the dependencies that we need to include, here are some use cases for the openfisca-**core** conda package ; feel free to correct/extend them 🙂: 

* A student in economics running Python code that calls an openfisca country model. 
This use case needs: 
  - a wheel of the country model 
  - the country model depend on a wheel of the openfisca-core repository ☸️ 
  - but this use cas doesn't need the openfisca-core web api or the dev dependencies.

* An administration official contributing to an openfisca country model on Windows OS. She/He might call openfisca Web API in production to answer the requests of a UI/simulator.
This use case needs:
  - the source code of the country model (from git clone) 
  - the country model will require an openfisca-core wheel ☸️.  
  - it might require openfisca-core web api dependencies 📲 to check that the changes made to the country model could be served by the web api without error. 
  - but this use case doesn't need openfisca-**core** dev dependencies (will need the country model dev dependencies).

* A user of the [CASD](https://www.casd.eu/en/) secured data hub that uses an openfisca country model and might need to test some code changes on the openfisca-core source code (to check performance or...). 
This use case needs:
  - a wheel of the country model and openfisca-core ☸️. 
  - when some code change is tested on openfisca-core, openfisca-core source code will be used instead of the wheel. Nevertheless, as the CASD is a closed environment where every request for a new library installation requires time, having dev 🧑‍💻 and even web api dependencies 📲 in the authorized openfisca-core library will be useful (for example, the user will gain time if `mypy`, `openfisca-country-template`... core dev dependencies are allowed through openfisca-core conda installation and will be able to run the tests on CASD machine).

Do you see any other use case? For example, what would be the use cas needing openfisca-core conda .tar.gz?

**Comment by benoit-cty** [comment]
Following my issue, the conda-forge team has opened an issue in conda-build for long filename : https://github.com/conda/conda-build/issues/4357, and helped me to find a workaround: https://github.com/conda-forge/staged-recipes/pull/17483#discussion_r788535778.

**Comment by benoit-cty** [comment]
To sum up we have:
- An OpenFisca-Core PR: https://github.com/openfisca/openfisca-core/pull/1105
- An OpenFisca-France PR: https://github.com/openfisca/openfisca-france/pull/1786
- A new repo for OpenFisca-Core in conda-forge : https://anaconda.org/conda-forge/openfisca-core
- A new repo for OpenFisca-Core and OpenFisca-France in an openfisca conda channel: https://anaconda.org/openfisca
Why 2 channels ? Because conda-forge seems better but more difficult to upgrade. So we will probably having less release on conda-forge channel, where openfisca channel will be kept in sync with PyPi automatically.
- A new github repo to update conda-forge : https://github.com/openfisca/openfisca-core-feedstock to submit PR to https://github.com/conda-forge/openfisca-core-feedstock
- A new github repo to add new project to conda-forge : https://github.com/openfisca/openfisca-conda-forge-recipes
- A PR waiting to add OpenFisca-France to conda-forge : https://github.com/conda-forge/staged-recipes/pull/17483
- An issue in conda-build to handle OpenFisca-France long file name in CI https://github.com/conda/conda-build/issues/4357

I wasn't expecting so much work 😪 But adding new repo of OpenFisca to conda will be much faster now.
